libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/components/BlurHashAsyncImage.kt

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (c) 2023 New Vector Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.element.android.libraries.designsystem.components
+
+import androidx.compose.animation.AnimatedVisibility
+import androidx.compose.animation.fadeIn
+import androidx.compose.animation.fadeOut
+import androidx.compose.foundation.Image
+import androidx.compose.foundation.layout.Box
+import androidx.compose.foundation.layout.fillMaxSize
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.DisposableEffect
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.saveable.rememberSaveable
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.asImageBitmap
+import androidx.compose.ui.layout.ContentScale
+import coil.compose.AsyncImage
+import com.vanniktech.blurhash.BlurHash
+
+@Composable
+fun BlurHashAsyncImage(
```

Comment:
This can be simplified down to something like:

```
@Composable
fun BlurHashAsyncImage(
    model: Any?,
    blurHash: String?,
    contentDescription: String?,
    modifier: Modifier = Modifier,
    contentScale: ContentScale = ContentScale.Fit,
) {
    AsyncImage(
        model = ImageRequest.Builder(LocalContext.current)
            .data(model)
            .crossfade(true)
            .build(),
        contentDescription = contentDescription,
        modifier = modifier,
        placeholder = blurHash?.let { hash ->
            BlurHash.decode(hash, 10, 10)?.asImageBitmap()?.let(::BitmapPainter)
        },
        contentScale = contentScale,
    )
}
```

Previews also work with this impl so we can add one.
What d'you think?

Or perhaps do you think this is too trivial and we can just create an utility function to build the Painter from a blurhash string?
Something like:
```
fun blurHashPainter(blurHash: String?): Painter? = blurHash?.let { hash ->
    BlurHash.decode(hash, 10, 10)?.asImageBitmap()?.let(::BitmapPainter)
}
```
And then use it with Coil's AsyncImage?

Reply:
Actually, I just moved this class to the design system module from a feature one since it's needed in several places. I don't mind simplifying the implementation if it has a similar UX.

Reply:
> Actually, I just moved this class to the design system module from a feature one since it's needed in several places. I don't mind simplifying the implementation if it has a similar UX.

The preview looks the same, I didn't check the crossfade animation though.

Reply:
There's one difference though, the original `BlurHashAsyncImage` applies `fillMaxSize()` to its inner AsyncImage.
We could do the same here but I just thought it'd be too opinionated for being a library composable.

Reply:
Anyhow I thought this was a new class, not a move, if you prefer I can do a separate rework PR once this is merged

Reply:
Yeah, that seems for the best. Also, the cross fade sometimes looks kind of weird, starting from a black square.

Reply:
> Is the Drawable being released from memory by the placeholder api?
Nope

> Also I think the painter should be remembered?
Yepp

Good thing Jorge didn't listen to me ;)

Reply:
@ganfra Why the need of calling `Bitmap.recycle()` though? Its doc string says it shouldn't normally be called.